src/sentry/dynamic_sampling/rules/utils.py

```diff
@@ -161,3 +164,22 @@ def get_enabled_user_biases(user_set_biases: Optional[List[ActivatableBias]]) ->
 
 def get_supported_biases_ids() -> Set[str]:
     return {bias["id"] for bias in DEFAULT_BIASES}
+
+
+def apply_dynamic_factor(base_sample_rate: float, x: float) -> float:
+    """
+    This function known as dynamic factor function is used during the rules generation in order to determine the factor
+    for each rule based on the base_sample_rate of the project.
+
+    The function decreases the factor from x to 1 in the interval [0.0, 1.0] which means that a user with
+    base_sample_rate = 0.0 will have a factor of x, whereas with base_sample_rate = 1.0 we will have a factor of 1.
```

Comment:
i think we will raise an exception if sample rate is 0.

Reply:
Yeah, that makes sense. It's not a problem mathematically but it doesn't make sense to compute any factors if that is the case.

Reply:
to demonstrate this you can add `assert base_sample_rate = 0.0`